File: survey/drivers/package_url/packageurl_ruby/format.rb
#!/usr/bin/env ruby

require 'json'

require_relative 'repo/lib/package_url'

STDIN.each_line {|line|
    line.chomp!
    return if line.empty?

    parts = JSON.parse(line)

    begin
        purl = PackageURL.new(
            type: parts["type"],
            name: parts["name"],
            namespace: parts["namespace"],
            version: parts["version"],
            qualifiers: parts["qualifiers"],
            subpath: parts["subpath"]
        )
        puts purl
    rescue => error
        puts JSON.generate({
            error: error.to_s,
        })
    end
    STDOUT.flush
}

File: survey/drivers/package_url/packageurl_js/format.mjs
#!/usr/bin/env node

import { PackageURL } from "packageurl-js";
import { createInterface } from "node:readline";

const rl = createInterface({
    input: process.stdin,
    output: process.stdout,
    terminal: false,
});

for await (let line of rl) {
    line = line.trim();
    if (line === "") {
        continue;
    }
    const parts = JSON.parse(line);
    try {
        const purl = new PackageURL(parts.type, parts.namespace, parts.name, parts.version, parts.qualifiers, parts.subpath);
        console.log(purl.toString());
    } catch (e) {
        console.log(JSON.stringify({"error": e.toString()}));
    }
}
